src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/v2/NpgStateService.kt

```diff
@@ -0,0 +1,125 @@
+import it.pagopa.ecommerce.commons.documents.v2.TransactionAuthorizationRequestedEvent
+import it.pagopa.ecommerce.commons.domain.v2.pojos.BaseTransaction
+import it.pagopa.ecommerce.commons.domain.v2.pojos.BaseTransactionWithPaymentToken
+import it.pagopa.ecommerce.commons.exceptions.NpgResponseException
+import it.pagopa.ecommerce.commons.generated.npg.v1.dto.StateResponseDto
+import it.pagopa.ecommerce.commons.queues.QueueEvent
+import it.pagopa.ecommerce.commons.queues.TracingInfo
+import it.pagopa.ecommerce.commons.utils.NpgPspApiKeysConfig
+import it.pagopa.ecommerce.eventdispatcher.exceptions.BadGatewayException
+import it.pagopa.ecommerce.eventdispatcher.exceptions.GetStateException
+import it.pagopa.ecommerce.eventdispatcher.exceptions.NoRetryAttemptsLeftException
+import it.pagopa.ecommerce.eventdispatcher.exceptions.TooLateRetryAttemptException
+import it.pagopa.ecommerce.eventdispatcher.queues.v2.QueueCommonsLogger.logger
+import java.time.Duration
+import java.time.Instant
+import java.util.*
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmpty
+
+@Component
+class NpgStateService(
+  @Autowired private val npgClient: NpgClient,
+  @Autowired private val npgCardsPspApiKey: NpgPspApiKeysConfig,
+  @Value("\${transactionAuthorizationRequested.paymentTokenValidityTimeOffset}")
+  private val paymentTokenValidityTimeOffset: Int,
+  @Autowired private val authRequestedQueueAsyncClient: QueueAsyncClient,
+  @Value("\${transactionAuthorizationRequested.eventOffsetSeconds}") private val retryOffset: Int,
+  @Value("\${transactionAuthorizationRequested.maxAttempts}") private val maxAttempts: Int,
+  @Value("\${azurestorage.queues.transientQueues.ttlSeconds}")
+  private val transientQueuesTTLSeconds: Int,
+) {
+
+  fun getStateNpg(transactionId: UUID, sessionId: String, pspId: String): Mono<StateResponseDto> {
+    return npgCardsPspApiKey[pspId].fold(
+      { ex -> Mono.error(ex) },
+      { apiKey ->
+        npgClient.getState(UUID.randomUUID(), sessionId, apiKey).onErrorMap(
+          NpgResponseException::class.java) { exception: NpgResponseException ->
+          val responseStatusCode = exception.statusCode
+          responseStatusCode
+            .map {
+              val errorCodeReason = "Received HTTP error code from NPG: $it"
+              if (it.is5xxServerError) {
+                BadGatewayException(errorCodeReason)
+              } else {
+                GetStateException(transactionId, errorCodeReason)
+              }
+            }
+            .orElse(GetStateException(transactionId, "Unknown NPG HTTP response code"))
+        }
+      })
+  }
+
+  fun enqueueRetryEvent(
+    baseTransaction: BaseTransaction,
+    retryCount: Int,
+    retryEvent: TransactionAuthorizationRequestedEvent,
```

Comment:
There is a main issue here. You are using the `TransactionAuthorizationRequestedEvent` as a retry event but there is no information about the retry count (seeing the above code retry count is always set to 0).
This will cause that all retries will be performed until payment token will be available having no control against the max retry attempts to be done.
Furthermore there will be no event into event store that traces the attempts made.  
Maybe a new retry event has to be add into commons library that will carry the retry count information

Reply:
To be addressed in future task